Convert all foreach() to range-based for()

thalieht opened the pull request:
...and fix coding style for various things.
I gave it a shot in converting a few "foreach" to "for" but i'm not sure i have grasped how qAsConst() and copyAsConst() should be used. From my understanding qAsConst() should be used for non-const qt containers and copyAsConst() for methods returning non-const qt containers? If i understood correctly i'll do the rest as well.

Another question is do you want me to change include guards to `#pragma once` in all header files?

LordNyriox commented:
@thalieht:  

>do you want me to change include guards to `#pragma once` in all header files? 

Personally, I would recommend doing so (in a separate commit, of course).  

I may be more of a power-user than a coder, but consistent implementation can only be a good thing, in my experience.  You can debug issues more easily that way.

LordNyriox commented on src/gui/properties/peerlistwidget.cpp:
```diff
@@ -102,7 +102,7 @@ PeerListWidget::PeerListWidget(PropertiesWidget *parent)
     // Ensure that at least one column is visible at all times
     bool atLeastOne = false;
-    for (unsigned int i = 0; i < PeerListDelegate::IP_HIDDEN; ++i) {
+    for (int i = 0; i < PeerListDelegate::IP_HIDDEN; ++i) {
```
@thalieht:  Speaking of me not being a coder… 

Why did you change this (and the two similar cases below) from an unsigned type to a signed one?  

I cannot imagine a negative width measurement being valid—which makes `unsigned int` look rather useful to me.

thalieht replied:
In this particular case there is no difference but there were other cases were `i` was converted implicitly to `int` anyway e.g. in line 115 and there was a warning. I just changed it here too for consistency.

LordNyriox replied:
@thalieht:  I figured it might be a conversion issue.  

Thanks for the clarification.

Chocobo1 commented on src/app/cmdoptions.cpp:
```diff
@@ -497,7 +498,7 @@ QString wrapText(const QString &text, int initialIndentation = USAGE_TEXT_COLUMN
     int currentLineMaxLength = wrapAtColumn - initialIndentation;
 
-    foreach (const QString &word, words.mid(1)) {
+    for (const QString &word : copyAsConst(words.mid(1))) {
```
I guess you won't need `copyAsConst()` here, also you could use [clazy](https://github.com/KDE/clazy) to help verify.

thalieht replied:
This is one of the 4 warnings Clazy complains.

glassez replied:
As for this change (and many other `copyAsConst` usages):
Since `mid()` doesn't actually return shared copy of some container `detach()` will not do anything except checking for references count so it's more lightweight than copy constructor called from `copyAsConst()`.

thalieht replied:
I have no idea how you know what you said and consequently, how to spot the other copyAsConst that aren't needed.

glassez replied:
As a rule, it obviously stems from the meaning of the function.
E.g. `mid()` here creates new list containing some subset of elements of affected list.
Just try to examine the meaning of all other affected functions. If you're in doubt in a particular case, ask for my explanation.

The problem of "detach" is greatly exaggerated. When we began to use range based for loop, we saw a warning that it can cause "detach()". And we blindly rushed to solve it, without going into details. But now it is obvious to me that in most cases this is not a problem at all.
@Chocobo1, don't you mind join this discussion?

thalieht replied:
There are a few existing copyAsConst that seem to return a new container e.g. https://github.com/qbittorrent/qBittorrent/blob/master/src/gui/ipsubnetwhitelistoptionsdialog.cpp#L49, https://github.com/qbittorrent/qBittorrent/blob/master/src/base/bittorrent/tracker.cpp#L139 and https://github.com/qbittorrent/qBittorrent/blob/master/src/webui/api/appcontroller.cpp#L205
Are they correct or do i remove them?
There is a case that returns a container that is returned from another function, it should keep copyAsConst right? https://github.com/qbittorrent/qBittorrent/blob/master/src/gui/properties/trackerlistwidget.cpp#L397

Chocobo1 replied:
If this isn't much work, it might be worth if we can make an non-exhaustive whitelist of situations (and put in coding guidelines) while reviewing this PR, specifying where `copyAsConst()` can be omitted. For example: `mid()` function in Qt types.

glassez replied:
>There are a few existing copyAsConst that seem to return a new container...
Are they correct or do i remove them?

It should be removed in all 3 cases above.
>There is a case that returns a container that is returned from another function, it should keep copyAsConst right?

No in case if first function returns no copy of existing object (as in your example).

In General, I came to the conclusion that the methods that return a newly created container (a copy of a temporary object, to be exact) are the vast majority. I do not even come to mind the opposite case (at least in Qt), it needs a good look. As for qBittorrent own code, the best solution would be to change the type of those few methods that return a shared copy of the class internal container so that they return a constant reference instead (e.g. https://github.com/qbittorrent/qBittorrent/pull/9883/files#diff-b369e532eff50253387c9d203da4b928R453) and remove `copyAsConst` at all so it did not tempt us. If we still encounter the affected method somewhere in the third-party code (or in Qt), we can always store its value in a pre-declared constant and iterate over it.

thalieht replied:
>we can always store its value in a pre-declared constant and iterate over it.

I thought copyAsConst was introduced to avoid that? https://github.com/qbittorrent/qBittorrent/pull/8325#discussion_r164265321

Chocobo1 commented on src/app/filelogger.cpp:
```diff
@@ -87,7 +88,7 @@ void FileLogger::deleteOld(const int age, const FileLogAgeType ageType)
     QDir dir(Utils::Fs::branchPath(m_path));
 
-    foreach (const QFileInfo file, dir.entryInfoList(QStringList("qbittorrent.log.bak*"), QDir::Files | QDir::Writable, QDir::Time | QDir::Reversed)) {
+    for (const QFileInfo &file : copyAsConst(dir.entryInfoList(QStringList("qbittorrent.log.bak*"), QDir::Files | QDir::Writable, QDir::Time | QDir::Reversed))) {
```
and here

thalieht replied:
This too.

Chocobo1 commented on src/base/bittorrent/magneturi.cpp:
```diff
@@ -83,10 +83,10 @@ MagnetUri::MagnetUri(const QString &source)
         m_trackers.append(QString::fromStdString(tracker));
 
-    foreach (const std::string &urlSeed, m_addTorrentParams.url_seeds)
+    for (const std::string &urlSeed : m_addTorrentParams.url_seeds)
```
seems need to add `qAsConst` for both loops.

thalieht replied:
But these are `std::vector` containers and this guy says:
>Avoiding Detaching
>
>If your container is a std:: container or QVarLengthArray, you replace the `Q_FOREACH(a, b)` with `for (a : b)` and you are done. Arguably, Q_FOREACH should never, ever have been used on such a container, since copying those always copies all elements (deep copy).

https://www.kdab.com/goodbye-q_foreach/

What am i missing?

Chocobo1 replied:
I didn't realize they are not Qt containers, so it's OK.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -198,7 +199,7 @@ namespace
         for (auto i = categories.cbegin(); i != categories.cend(); ++i) {
             const QString &category = i.key();
-            foreach (const QString &subcat, Session::expandCategory(category)) {
+            for (const QString &subcat : copyAsConst(Session::expandCategory(category))) {
```
seems no need here

Chocobo1 replied:
or you can just remove all the newly added `copyAsConst` and only add them back when clazy complains :P

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -774,12 +775,12 @@ bool Session::editCategory(const QString &name, const QString &savePath)
     m_storedCategories = map_cast(m_categories);
     if (isDisableAutoTMMWhenCategorySavePathChanged()) {
-        foreach (TorrentHandle *const torrent, torrents())
+        for (TorrentHandle *const torrent : qAsConst(m_torrents))
```
you should not replace the use of `torrents()` by `m_torrents`, those are different objects, or you should make another commit for it.

thalieht replied:
Oops this was a copy paste mistake i'll revert these.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1982,7 +1983,7 @@ void Session::increaseTorrentsPriority(const QStringList &hashes)
 
     // Sort torrents by priority
-    foreach (const InfoHash &hash, hashes) {
+    for (const InfoHash hash : hashes) {
```
why remove the `&`?

thalieht replied:
warning: loop variable 'hash' has type 'const BitTorrent::InfoHash &' but is initialized with type 'const QString' resulting in a copy
use non-reference type 'BitTorrent::InfoHash' to keep the copy or type 'const QString &' to prevent copying

Chocobo1 replied:
ok, do you mind rename `hash` -> `info` (or `infoHash`)? I find the old name easily mixed up with the element type of the container, and the ones below.

thalieht replied:
Certainly.

thalieht replied:
I did it in the coding style commit.

thalieht commented:
Clazy found only 4 in whole project, great. The condition i said for copyAsConst is obviously wrong but if all it takes is Clazy i guess it doesn't matter. But what about qAsConst? Is what i said in the first post correct or there is more?

thalieht commented:
I've also found a few existing `for ( : )` like the following
https://github.com/qbittorrent/qBittorrent/blob/547140ec4614d6e8597dd8217b928ffedcf000fc/src/base/rss/rss_feed.cpp#L232

`articles` is non-const QList. Is it correct as is, or it needs qAsConst?

Chocobo1 commented:
>But what about qAsConst? Is what i said in the first post correct or there is more?

Your understanding seems correct.

>articles is non-const QList. Is it correct as is, or it needs qAsConst?

Seems not as `result` is already `const`.

thalieht commented:
Thanks and sorry if i'm being a pain... i found the following from another project which is in the same situation:
https://github.com/lxqt/lxqt-config/pull/137/files#diff-e5a213b77fa1d61778446c5c0cb7dfa9R54
`info` being const but `variants` not, yet qAsConst is added anyway.

Chocobo1 commented on src/base/utils/string.h:
```diff
@@ -60,7 +60,7 @@ namespace Utils
             if (str.length() < 2) return str;
 
-            for (auto const quote : quotes) {
+            for (auto const &quote : quotes) {
```
adding & adds another indirection which is not necessary here, also you should switch `auto const` -> `const auto`

thalieht replied:
I had a warning here:
`warning: loop variable 'quote' of type 'const QChar' creates a copy from type 'const QChar'
use reference type 'const QChar &' to prevent copying`

Chocobo1 replied:
ok, just switch the order of "auto" & "const".

Chocobo1 commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -373,7 +373,7 @@ void AutoDownloader::addJobForArticle(Article *article)
 void AutoDownloader::processJob(const QSharedPointer<ProcessingJob> &job)
 {
-    for (AutoDownloadRule &rule : m_rules) {
+    for (AutoDownloadRule rule : qAsConst(m_rules)) {
```
can it bee `const AutoDownloadRule &rule`?

thalieht replied:
No... 3 lines below for const:
`error: 'this' argument to member function 'accepts' has type 'const RSS::AutoDownloadRule', but function is not marked const`
and 1 line below for reference:
`error: binding value of type 'const RSS::AutoDownloadRule' to reference to type 'RSS::AutoDownloadRule' drops 'const' qualifier`

I was wondering what's preferable here, keep the reference or use qAsConst?

Chocobo1 replied:
>I was wondering what's preferable here, keep the reference or use qAsConst?

I would avoid doing changes here, it is not legacy code.

thalieht replied:
Actually this line is more than 1.5 years old, before qAsConst was introduced to qBt
https://github.com/qbittorrent/qBittorrent/commit/989a70fe60bb28c5c6bfdce86d6bec5949a833f0#diff-576c6aa10c20e1e0d491fd5cf5b81426R234
Maybe @glassez can say if it should be changed.

glassez replied:
This change is incorrect. Actual m_rules items should be modified inside this loop.

Chocobo1 commented on src/base/bittorrent/torrenthandle.cpp:
```diff
@@ -400,13 +400,13 @@ void TorrentHandle::addTrackers(const QList<TrackerEntry> &trackers)
 }
 
-void TorrentHandle::replaceTrackers(QList<TrackerEntry> trackers)
+void TorrentHandle::replaceTrackers(const QList<TrackerEntry> trackers)
```
please try `const QList<TrackerEntry> &trackers`

Chocobo1 commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -451,7 +451,7 @@ void AutoDownloader::store()
 
     QJsonObject jsonObj;
-    foreach (auto rule, m_rules)
+    for (auto &rule : qAsConst(m_rules))
```
`const auto &rule`?

thalieht replied:
I thought i had tried that but guess not, thanks it works.

Chocobo1 commented on src/base/utils/misc.cpp:
```diff
@@ -391,23 +391,23 @@ QString Utils::Misc::getUserIDString()
 {
     QStringList ret;
-    foreach (const bool &b, l)
+    for (const bool &b : l)
```
please also change to `const bool b`

thalieht replied:
ok

Chocobo1 commented on src/gui/properties/trackersadditiondialog.cpp:
```diff
@@ -59,7 +59,7 @@ TrackersAdditionDialog::~TrackersAdditionDialog()
 {
     QStringList cleanTrackers;
-    foreach (QString url, m_ui->textEditTrackersList->toPlainText().split('\n')) {
+    for (QString &url : m_ui->textEditTrackersList->toPlainText().split('\n')) {
```
this should still be `QString url`

thalieht replied:
ok

Chocobo1 commented on src/base/bittorrent/torrenthandle.cpp:
```diff
@@ -596,8 +596,7 @@ bool TorrentHandle::removeTag(const QString &tag)
 {
-    // QT automatically copies the container in foreach, so it's safe to mutate it.
-    foreach (const QString &tag, m_tags)
+    for (const QString &tag : qAsConst(m_tags))
```
I think this form will have problems (crash), better:
```c++
const QSet<QString> tags {m_tags};
for (const QString &tag : tags)
    removeTag(tag);
```

Chocobo1 replied:
or
```c++
for (const QString &tag : tags())
    removeTag(tag);
```

thalieht replied:
Actually i have tested qAsConst and what you said and it doesn't crash, also tested without any of them and naturally it crashed. If you still think i should do it i will.

thalieht replied:
Never mind it crashed when doing it to multiple torrents.

thalieht commented on src/webui/webapplication.cpp:
```diff
@@ -645,7 +645,7 @@ void WebApplication::sessionStart()
     // remove outdated sessions
     const qint64 now = QDateTime::currentMSecsSinceEpoch() / 1000;
-    foreach (const auto session, m_sessions) {
+    for (const auto session : qAsConst(m_sessions)) {
```
I looked at all of them again and i only found this one. I'll make a copy of `m_sessions`.

thalieht commented on src/base/bittorrent/session.cpp:
```diff
@@ -1780,7 +1781,7 @@ void Session::processShareLimits()
     qDebug("Processing share limits...");
 
-    foreach (TorrentHandle *const torrent, m_torrents) {
+    for (TorrentHandle *const torrent : qAsConst(m_torrents)) {
```
I found this too, `deleteTorrent()` modifies m_torrents, i'll change `m_torrents` to `torrents()`

Chocobo1 commented:
Nothing more from me, I think you can squash commits then let others review.

thalieht commented:
Squashed.

glassez commented:
Sorry for delay.
Guys, let's define the policy of using qAsConst/copyAsConst once and for all.
1. **(lvalue)** We require to use qAsConst when we iterate over non-const container but we don't intend to modify its items. However we can safely use it even with const containers to be sure that further code modifications will not lead to unexpected detaching of shared data.
2. **(rvalue)** copyAsConst is a more expensive operation because it calls the copy constructor (although it is quite lightweight). When can it really help? Only when affected method returns copy of internal container. When method creates new container it isn't affected by "detaching" problem. Now I'm not sure that using copyAsConst is the best solution. It generates either unnecessary copying or requires awareness of the internals of the affected methods (i.e. we need to know exactly what it returns to decide whether to use copyAsConst or not). Maybe just drop it and use another approach?

thalieht commented:
I added qAsConst if container is not const or where possible and copyAsConst where Clazy said it might detach. If not copyAsConst then what? Also i'm not sure about https://github.com/qbittorrent/qBittorrent/pull/9824#issuecomment-437591934 and https://github.com/qbittorrent/qBittorrent/pull/9824#discussion_r232946145

thalieht commented:
I run Clazy again and i had missed one, fixed.

thalieht commented:
Reverted and thanks for the link.

thalieht commented:
Removed most copyAsConst. The remaining ones are mostly qt functions that i have no idea whats going on inside:
[QNetworkCookieJar::allCookies()](https://github.com/qt/qtbase/blob/c5307203f5c0b0e588cc93e70764c090dd4c2ce0/src/network/access/qnetworkcookiejar.cpp#L121)
		[QSettings::allKeys()](https://github.com/qt/qtbase/blob/527406cbd99f44470ef87468b73c18df949e8ac7/src/corelib/io/qsettings.cpp#L3133)
		[QItemSelectionModel::selectedIndexes()](https://github.com/qt/qtbase/blob/c5307203f5c0b0e588cc93e70764c090dd4c2ce0/src/corelib/itemmodels/qitemselectionmodel.cpp#L1703)
		[QApplication::allWidgets()](https://github.com/qt/qtbase/blob/825f98815683faea06144ab0262129b0367798ee/src/widgets/kernel/qapplication.cpp#L1736)
		[QWidget::actions()](https://github.com/qt/qtbase/blob/c5307203f5c0b0e588cc93e70764c090dd4c2ce0/src/widgets/kernel/qwidget.cpp#L3354)
		QObject::findChildren<T>()
		[QButtonGroup::buttons()](https://github.com/qt/qtbase/blob/c5307203f5c0b0e588cc93e70764c090dd4c2ce0/src/widgets/widgets/qbuttongroup.cpp#L287)
		https://github.com/qbittorrent/qBittorrent/blob/0e8323659e026b72d13d40f942e6317c9b2591f4/src/gui/categoryfilterwidget.cpp#L234

Also there is `Feed::articles()` which returns shared container, and `Folder::articles()` which does not. The following is confusing to me and i'm not sure which one is used here (i assume `Feed::articles()`)
https://github.com/qbittorrent/qBittorrent/blob/0e8323659e026b72d13d40f942e6317c9b2591f4/src/gui/rss/articlelistwidget.cpp#L71

I can't make `Feed::articles()` return const ref, i get some error about inheritance.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -845,7 +845,7 @@ void Session::setSubcategoriesEnabled(bool value)
 }
 
-QSet<QString> Session::tags() const
+const QSet<QString> &Session::tags() const
```
@thalieht 
Now I suspect the [problem](https://github.com/qbittorrent/qBittorrent/pull/9824#discussion_r232918799) returns.

Also ping @glassez, the same apply to your change at https://github.com/qbittorrent/qBittorrent/pull/9883/files#diff-b369e532eff50253387c9d203da4b928R453

glassez replied:
We shouldn't modify containers inside range based `for` loop! In such case we should either iterate over its copy or use another way.

Chocobo1 replied:
Now I suspect "Return const reference for a few shared containers" is not going the right way, it changes the whole semantics of the function... aren't we used to getting a new copy of container from Qt functions? Now we only get constant references, it is a big difference...

glassez replied:
>aren't we used to getting a new copy of container from Qt functions? Now we only get constant references, it is a big difference...

Damn. You're right.
So what are we supposed to do? There are two main options:
1. Apply `copyAsConst()` in a "paranoid way" to all rvalues.
2. Apply `copyAsConst()` only to the rvalues that produces shared copy of inner containers (if this follows from the meaning of such method, because we should not get into its implementation).

thalieht replied:
Will delete this commit.

Chocobo1 commented:
>The remaining ones are mostly qt functions that i have no idea whats going on inside:

This question makes me thinking, normally we should only care about the return type of a function, not its implementation, however the detachment problem forces us to look into, isn't that a bit weird?
Another (less popular) idea is just ignore using `copyAsConst` in most cases, only use it at performance critical places (almost none in qbt I guess), does that make sense?

glassez commented:
>Another (less popular) idea is just ignore using copyAsConst in most cases, only use it at performance critical places (almost none in qbt I guess), does that make sense?

I wouldn't go that way. After all, the copy constructor is less expensive than the real `detach()` (which leads to deep copying). So, if I had to choose between extreme cases, I would prefer to use `copyAsConst()`. In addition, it will protect against changing the program logic by oversight (because `for (Type var : copyAsConst(container))` is equivalent to `foreach (Type var, container)`).
So the only question is whether we should "lazily" use it everywhere, or try to discard those cases where it is not required.

thalieht commented:
I think we should keep it for Qt functions because we don't know if they will be changed some day to return shared container, except maybe for ones that wouldn't make sense to, e.g. mid() split().
Awaiting your decision guys.

glassez commented:
I had to re-examine it...
**First.** Why did we call it `copyAsConst`? It totally baffled me! Let's see its code:
```c++
template <typename T>
constexpr typename std::add_const<T>::type copyAsConst(T &&t) noexcept { return std::move(t); }
```
We don't really copy the value but move it. So the better name should be "forwardAsConst" or "fwdAsConst" (or something like it).
**Second.** Move-constructor is really simple. It just copies a couple of pointers.
**Third." Let's compare the pros and cons. In the case when we are dealing with a really shared container (references count > 1), then `copyAsConst` wins undoubtedly. When a container exists in a single instance (I mean its shard data), we get the following:
1. `copyAsConst` way causes Move-constructor + destructor of empty object (so lightweight)
2. non-`copyAsConst` causes two `detach()` (in its lightweight form, i.e. just checking ref count)
Additionally the first way prevents us from unnecessary work to assess the need for use `copyAsConst` in some particular case (and from errors when we misinterpret some case).

I'm sorry I made you think of that. But I think we needed to do that before we approved that PR.
So let's choose our way and finish it!

Chocobo1 commented:
>So the only question is whether we should "lazily" use it everywhere, or try to discard those cases where it is not required.

Personally I have no problem using `copyAsConst` everywhere, except the name is a bit long and too verbose which slows me down when reading it, sometimes I hope we can have it in some shorthand, e.g. `fConst()` which stands for forward as const.

glassez commented:
>shorthand, e.g. fConst() which stands for forward as const.

I would use more readable name, e.g. `asConst()`. We can even rename our implementation of qAsConst to asConst to use single name in all logically similar situations (for both lvalues and rvalues).

Chocobo1 commented:
>I would use more readable name, e.g. asConst(). We can even rename our implementation of qAsConst to asConst to use single name in all logically similar situations (for both lvalues and rvalues).

Agree, also I just remembered that it was proposed as `constify()` in PR #8325 but was changed later.

glassez commented:
>Agree, also I just remembered that it was proposed as `constify()` in PR #8325 but was changed later.

Well, let it be `constify()` if you prefer it more.
IIRC, you don't object to have same name for them?

Chocobo1 commented:
>IIRC, you don't object to have same name for them?

No problem for me, actually `asConst()` is 1 char shorter than the other so lets just use it.

thalieht commented:
So i basically remove the last 2 commits and then rename copyAsConst and qAsConst to asConst in a new commit? Wasn't qAsConst chosen so qBt can use Qt's implementation after min version is bumped to >= 5.7.0?

glassez commented:
>Wasn't qAsConst chosen so qBt can use Qt's implementation after min version is bumped to >= 5.7.0?

It's just trivial function template so we won't win much by using Qt's qAsConst instead of our own implementation.

thalieht commented:
Removed last 2 commits (kept a backup just in case) and renamed qAsConst() and copyAsConst() to asConst() so as far as copyAsConst goes we're back to what Clazy complained might detach.

thalieht commented on src/base/global.h:
```diff
@@ -34,15 +34,13 @@
 template <typename T>
-void qAsConst(const T &&) = delete;
-#endif
+void asConst(const T &&) = delete;
```
I'm not entirely sure what's going on here to be honest. Is this still needed?

glassez replied:
I would move this deleted template to the end.

thalieht replied:
Done.

glassez commented on src/base/global.h:
```diff
@@ -34,15 +34,13 @@
 const char C_TORRENT_FILE_EXTENSION[] = ".torrent";
 
 
-#if QT_VERSION < QT_VERSION_CHECK(5, 7, 0)
```
Please remove also redundant blank line above.

thalieht replied:
Done.

glassez commented on src/base/global.h:
```diff
@@ -34,15 +34,13 @@
-#endif
+void asConst(const T &&) = delete;
 
 // returns a const object copy
```
It's better to say something like "Forward rvalue as const". We don't really copy it.

thalieht replied:
Done.

glassez commented on src/base/global.h:
```diff
@@ -34,15 +34,13 @@
-constexpr typename std::add_const<T>::type &qAsConst(T &t) noexcept { return t; }
+constexpr typename std::add_const<T>::type &asConst(T &t) noexcept { return t; }
 
 // prevent rvalue arguments:
```
This comment has became incorrect now. We don't really prevent rvalues now. This deleted template actually prevents only arguments declared as `const Container func(...)`.

glassez replied:
"prevent const rvalue arguments" is more correct, IIRC.

thalieht replied:
Done.

thalieht commented:
Rebased.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -775,12 +775,12 @@ bool Session::editCategory(const QString &name, const QString &savePath)
     m_storedCategories = map_cast(m_categories);
     if (isDisableAutoTMMWhenCategorySavePathChanged()) {
-        foreach (TorrentHandle *const torrent, torrents())
+        for (TorrentHandle *const torrent : torrents())
```
Why didn't you apply `asConst()` to all `torrents()` calls?

thalieht replied:
Clazy doesn't complain about possible detachment here.

glassez replied:
>Clazy doesn't complain about possible detachment on them.

But it REALLY does! And it is REAL detachment in this case (unlike of `mid()` which Clazy complains about).
Clazy gives too many false positives and incorrect results for us to blindly believe it.

thalieht replied:
Oh no... there are other places i haven't added asConst. Do i just add it everywhere then?

glassez replied:
>whether i should add asConst to the ones that don't need it.

If you can identify them for sure...
Adding asConst to the ones that don't need it is less evil than skipping the ones that are prone to the detaching problem.

thalieht replied:
I can, as you said if the container is created in the function then it's not shared. But if the function changes, the for() loops will surely be forgotten.

Chocobo1 replied:
>Adding asConst to the ones that don't need it is less evil than skipping the ones that are prone to the detaching problem.

Agree.

thalieht commented on src/base/bittorrent/session.cpp:
```diff
@@ -1982,8 +1983,8 @@ void Session::increaseTorrentsPriority(const QStringList &hashes)
     // Sort torrents by priority
-    foreach (const InfoHash &hash, hashes) {
-        TorrentHandle *const torrent = m_torrents.value(hash);
+    for (const InfoHash infoHash : hashes) {
```
https://github.com/qbittorrent/qBittorrent/pull/9824#discussion_r232156976

thalieht commented:
Added asConst to all functions. I'm open to changes if you want some specific ones removed.

Chocobo1 commented:
Thank you!
